File: src/libs/base/arch/gem5/Env.cc
/*
 * Copyright (C) 2016, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/Common.h>
#include <base/stream/Serial.h>
#include <base/CPU.h>
#include <base/Env.h>

#include <exception>
#include <functional>
#include <stdlib.h>

typedef void (*constr_func)();

extern constr_func CTORS_BEGIN;
extern constr_func CTORS_END;

EXTERN_C void __cxa_finalize(void *);
EXTERN_C void _init();
EXTERN_C void init_env(m3::Env *env);
EXTERN_C int main(int argc, char **argv);

namespace m3 {

OStream &operator<<(OStream &os, const Env &senv) {
    os << "pe_id        : " << senv.pe_id << "\n";
    os << "pe_desc      : " << fmt(senv.pe_desc, "#x") << "\n";
    os << "argc         : " << senv.argc << "\n";
    os << "argv         : " << fmt(senv.argv, "p") << "\n";
    os << "heap_size    : " << fmt(senv.heap_size, "#x") << "\n";
    os << "pe_mem_base  : " << fmt(senv.pe_mem_base, "p") << "\n";
    os << "pe_mem_size  : " << fmt(senv.pe_mem_size, "#x") << "\n";
    os << "sp           : " << fmt(senv.sp, "p") << "\n";
    os << "entry        : " << fmt(senv.entry, "p") << "\n";
    os << "first_std_ep : " << senv.first_std_ep << "\n";
    os << "first_sel    : " << senv.first_sel << "\n";
    os << "lambda       : " << fmt(senv.lambda, "p") << "\n";
    os << "rmng_sel     : " << senv.rmng_sel << "\n";
    os << "pager_sess   : " << senv.pager_sess << "\n";
    os << "mounts_addr  : " << fmt(senv.mounts_addr, "p") << "\n";
    os << "mounts_len   : " << senv.mounts_len << "\n";
    os << "fds_addr     : " << senv.fds_addr << "\n";
    os << "fds_len      : " << fmt(senv.fds_len, "p") << "\n";
    os << "rbuf_cur     : " << fmt(senv.rbuf_cur, "p") << "\n";
    os << "rbuf_end     : " << fmt(senv.rbuf_end, "p") << "\n";
    os << "backend_addr : " << fmt(senv.backend_addr, "p") << "\n";
    return os;
}

void Env::pre_init() {
}

void Env::post_init() {
    // call constructors
    _init();
    for(constr_func *func = &CTORS_BEGIN; func < &CTORS_END; ++func)
        (*func)();
}

void Env::pre_exit() {
}

void Env::run() {
    Env *e = env();

    int res;
    if(e->lambda) {
        e->backend()->reinit();

        std::function<int()> *f = reinterpret_cast<std::function<int()>*>(e->lambda);
        res = (*f)();
    }
    else {
        init_env(e);
        e->pre_init();
        e->backend()->init();
        e->post_init();

        char **argv = reinterpret_cast<char**>(e->argv);
        if(sizeof(char*) != sizeof(uint64_t)) {
            uint64_t *argv64 = reinterpret_cast<uint64_t*>(e->argv);
            argv = new char*[e->argc];
            for(uint64_t i = 0; i < e->argc; ++i)
                argv[i] = reinterpret_cast<char*>(argv64[i]);
        }
        res = main(static_cast<int>(e->argc), argv);
    }

    e->exit(res, false);
    UNREACHED;
}

USED void Env::exit(int code, bool abort) {
    pre_exit();
    if(!abort)
        __cxa_finalize(nullptr);
    backend()->exit(code);
    UNREACHED;
}

}

File: src/libs/rust/m3/src/arch/kachel/env.rs
/*
 * Copyright (C) 2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use base;
use core::cmp;
use core::intrinsics;

use crate::cap::Selector;
use crate::cfg;
use crate::com::SendGate;
use crate::env;
use crate::kif::{self, PEDesc};
use crate::pes::VPE;
use crate::serialize::Source;
use crate::session::{Pager, ResMng};
use crate::tcu;
use crate::util;
use crate::vfs::{FileTable, MountTable};

#[derive(Default, Copy, Clone)]
#[repr(C)]
pub struct EnvData {
    base: base::envdata::EnvData,
}

impl EnvData {
    pub fn platform(&self) -> base::envdata::Platform {
        base::envdata::Platform::from(self.base.platform)
    }

    pub fn set_platform(&mut self, platform: base::envdata::Platform) {
        self.base.platform = platform.val as u64
    }

    pub fn pe_id(&self) -> u64 {
        self.base.pe_id
    }

    pub fn shared(&self) -> bool {
        self.base.shared != 0
    }

    pub fn pe_desc(&self) -> PEDesc {
        PEDesc::new_from(self.base.pe_desc)
    }

    pub fn set_pedesc(&mut self, pe: PEDesc) {
        self.base.pe_desc = pe.value();
    }

    pub fn argc(&self) -> usize {
        self.base.argc as usize
    }

    pub fn set_argc(&mut self, argc: usize) {
        self.base.argc = argc as u64;
    }

    pub fn set_argv(&mut self, argv: usize) {
        self.base.argv = argv as u64;
    }

    pub fn sp(&self) -> usize {
        self.base.sp as usize
    }

    pub fn set_sp(&mut self, sp: usize) {
        self.base.sp = sp as u64;
    }

    pub fn set_entry(&mut self, entry: usize) {
        self.base.entry = entry as u64;
    }

    pub fn heap_size(&self) -> usize {
        self.base.heap_size as usize
    }

    pub fn set_heap_size(&mut self, size: usize) {
        self.base.heap_size = size as u64;
    }

    pub fn first_std_ep(&self) -> tcu::EpId {
        self.base.first_std_ep as tcu::EpId
    }

    pub fn set_first_std_ep(&mut self, start: tcu::EpId) {
        self.base.first_std_ep = start as u64;
    }

    pub fn has_vpe(&self) -> bool {
        self.base.vpe_addr != 0
    }

    pub fn vpe(&self) -> &'static mut VPE {
        // safety: we trust our loader
        unsafe { intrinsics::transmute(self.base.vpe_addr as usize) }
    }

    pub fn load_pager(&self) -> Option<Pager> {
        match self.base.pager_sess {
            0 => None,
            s => Some(Pager::new_bind(s as Selector).unwrap()),
        }
    }

    pub fn load_rmng(&self) -> Option<ResMng> {
        match self.base.rmng_sel as Selector {
            kif::INVALID_SEL => None,
            s => Some(ResMng::new(SendGate::new_bind(s))),
        }
    }

    pub fn load_first_sel(&self) -> Selector {
        // it's initially 0. make sure it's at least the first usable selector
        cmp::max(kif::FIRST_FREE_SEL, self.base.first_sel as Selector)
    }

    pub fn load_mounts(&self) -> MountTable {
        if self.base.mounts_len != 0 {
            // safety: we trust our loader
            let slice = unsafe {
                util::slice_for(
                    self.base.mounts_addr as *const u64,
                    self.base.mounts_len as usize,
                )
            };
            MountTable::unserialize(&mut Source::new(slice))
        }
        else {
            MountTable::default()
        }
    }

    pub fn load_fds(&self) -> FileTable {
        if self.base.fds_len != 0 {
            // safety: we trust our loader
            let slice = unsafe {
                util::slice_for(self.base.fds_addr as *const u64, self.base.fds_len as usize)
            };
            FileTable::unserialize(&mut Source::new(slice))
        }
        else {
            FileTable::default()
        }
    }

    // --- gem5 specific API ---

    pub fn set_vpe(&mut self, vpe: &VPE) {
        self.base.vpe_addr = vpe as *const VPE as u64;
    }

    pub fn has_lambda(&self) -> bool {
        self.base.lambda == 1
    }

    pub fn set_lambda(&mut self, lambda: bool) {
        self.base.lambda = lambda as u64;
    }

    pub fn set_first_sel(&mut self, sel: Selector) {
        self.base.first_sel = sel;
    }

    pub fn set_rmng(&mut self, sel: Selector) {
        self.base.rmng_sel = sel;
    }

    pub fn set_files(&mut self, off: usize, len: usize) {
        self.base.fds_addr = off as u64;
        self.base.fds_len = len as u64;
    }

    pub fn set_mounts(&mut self, off: usize, len: usize) {
        self.base.mounts_addr = off as u64;
        self.base.mounts_len = len as u64;
    }

    pub fn set_pager(&mut self, pager: &Pager) {
        self.base.pager_sess = pager.sel() as u64;
    }
}

pub fn get() -> &'static mut EnvData {
    // safety: we trust our loader
    unsafe { intrinsics::transmute(cfg::ENV_START) }
}

pub fn closure() -> &'static mut env::Closure {
    // safety: we trust our loader
    unsafe { intrinsics::transmute(cfg::ENV_START + util::size_of::<EnvData>()) }
}
